File: iOS/CatchTheRule/Features/Play/FigureView.swift
import SwiftUI

// MARK: - 도형 규칙 렌더러
//
// 두 가지 표현을 그린다.
//  - 숫자형(figure.hasSlots): 도형 윤곽 + 꼭짓점/세그먼트/중앙에 숫자. nil 슬롯이 빈칸.
//  - 시각형(no slots): rotation/filled/count 로 순수 모양을 그린다(시퀀스/보기 공용).

// MARK: 숫자형 — 예시 도형 여러 개를 한 줄에

/// 숫자형 퍼즐: 완성된 예시 도형 + 빈칸 도형을 나란히 보여줘 규칙을 추론하게 한다.
struct FigureNumberRow: View {
    let figures: [Figure]
    var blankText: String = "?"
    var feedback: AnswerFeedback? = nil

    var body: some View {
        // 도형 수에 따라 크기 축소(가로 넘침 방지).
        let n = max(1, figures.count)
        let side: CGFloat = n >= 3 ? 100 : (n == 2 ? 138 : 150)
        let spacing: CGFloat = n >= 3 ? 6 : 12
        HStack(spacing: spacing) {
            ForEach(Array(figures.enumerated()), id: \.offset) { _, fig in
                FigureNumberView(figure: fig, blankText: blankText, feedback: feedback, side: side)
            }
        }
        .frame(maxWidth: .infinity)
    }
}

/// 숫자형 도형 1개(삼각형/사각형/원). 빈칸 슬롯은 blankText 로 표시.
struct FigureNumberView: View {
    let figure: Figure
    var blankText: String = "?"
    var feedback: AnswerFeedback? = nil
    var side: CGFloat = 150

    private var slots: [String?] { figure.slots ?? [] }
    private var labelSide: CGFloat { max(28, side * 0.27) }
    private var labelFont: CGFloat { max(14, side * 0.135) }

    var body: some View {
        ZStack {
            outline
            ForEach(Array(anchors().enumerated()), id: \.offset) { i, pt in
                slotLabel(slots[safe: i] ?? "", at: pt)
            }
        }
        .frame(width: side, height: side)
    }

    // 도형 윤곽선
    private var outline: some View {
        let inset = side * 0.16
        return ZStack {
            switch figure.shape {
            case "square":
                RoundedRectangle(cornerRadius: 10, style: .continuous)
                    .stroke(Theme.stroke, lineWidth: 2)
                    .padding(inset)
            case "circle":
                Circle().stroke(Theme.stroke, lineWidth: 2).padding(inset)
                dividers   // 세그먼트 구분선
            default: // triangle
                TriangleShape().stroke(Theme.stroke, style: StrokeStyle(lineWidth: 2, lineJoin: .round))
                    .padding(inset)
            }
        }
    }

    // 원 세그먼트 구분선(슬롯 경계마다 중심→가장자리).
    private var dividers: some View {
        let c = CGPoint(x: side / 2, y: side / 2)
        let edges: [CGPoint] = dividerEdges()
        return ZStack {
            ForEach(Array(edges.enumerated()), id: \.offset) { _, edge in
                Path { p in
                    p.move(to: c)
                    p.addLine(to: edge)
                }
                .stroke(Theme.stroke.opacity(0.6), lineWidth: 1)
            }
        }
    }

    private func dividerEdges() -> [CGPoint] {
        let n = max(1, slots.count)
        let cx = Double(side) / 2, cy = Double(side) / 2
        let r = Double(side) / 2 - Double(side) * 0.16
        let count = Double(n)
        return (0..<n).map { i -> CGPoint in
            let deg: Double = -90.0 + (Double(i) + 0.5) * 360.0 / count
            let ang: Double = deg * .pi / 180
            return CGPoint(x: cx + cos(ang) * r, y: cy + sin(ang) * r)
        }
    }

    // 슬롯 숫자(또는 빈칸).
    @ViewBuilder
    private func slotLabel(_ value: String?, at pt: CGPoint) -> some View {
        let isBlank = (value == nil)
        let text = isBlank ? blankText : (value ?? "")
        Text(text)
            .font(.system(size: labelFont, weight: .bold, design: .rounded))
            .minimumScaleFactor(0.5)
            .lineLimit(1)
            .foregroundStyle(isBlank ? Theme.textPrimary : Theme.textSecondary)
            .frame(width: labelSide, height: labelSide)
            .background(
                Circle().fill(isBlank ? Color.white.opacity(0.05) : Theme.card)
            )
            .overlay(
                Circle().stroke(isBlank ? blankStroke : AnyShapeStyle(Theme.stroke),
                                lineWidth: isBlank ? 2 : 1)
            )
            .shadow(color: Theme.success.opacity(isBlank && feedback == .correct ? 0.7 : 0), radius: 10)
            .position(pt)
    }

    private var blankStroke: AnyShapeStyle {
        switch feedback {
        case .correct: return AnyShapeStyle(Theme.success)
        case .wrong: return AnyShapeStyle(Theme.danger)
        case .none: return AnyShapeStyle(Theme.accentGradient)
        }
    }

    // 슬롯 앵커 좌표(시계방향, 위에서 시작). 마지막 1칸이 남으면 중앙(triangle/square).
    private func anchors() -> [CGPoint] {
        let n = slots.count
        let inset = side * 0.16
        let center = CGPoint(x: side / 2, y: side / 2)
        switch figure.shape {
        case "square":
            let lo = inset, hi = side - inset
            let corners = [CGPoint(x: lo, y: lo), CGPoint(x: hi, y: lo),
                           CGPoint(x: hi, y: hi), CGPoint(x: lo, y: hi)]
            return n > 4 ? corners + [center] : Array(corners.prefix(n))
        case "circle":
            let cx = Double(center.x), cy = Double(center.y)
            let r = Double(side) / 2 - Double(side) * 0.3
            let count = Double(max(1, n))
            return (0..<n).map { i -> CGPoint in
                let ang: Double = (-90.0 + Double(i) * 360.0 / count) * .pi / 180
                return CGPoint(x: cx + cos(ang) * r, y: cy + sin(ang) * r)
            }
        default: // triangle
            let lo = inset, hi = side - inset
            // 라벨이 변에 겹치지 않도록 꼭짓점을 살짝 안쪽으로.
            let verts = [CGPoint(x: side / 2, y: lo + 6),
                         CGPoint(x: hi - 6, y: hi - 4),
                         CGPoint(x: lo + 6, y: hi - 4)]
            // 삼각형 무게중심은 박스 정중앙보다 아래 → 중앙 슬롯을 꼭짓점 평균으로.
            let triCenter = CGPoint(x: side / 2,
                                    y: verts.reduce(0) { $0 + $1.y } / CGFloat(verts.count))
            return n > 3 ? verts + [triCenter] : Array(verts.prefix(n))
        }
    }
}

// MARK: 시각형 — 순수 모양

/// 시각 규칙용 도형(회전/채움/개수). 시퀀스 셀·보기 공용.
struct FigureGlyph: View {
    let figure: Figure
    var size: CGFloat = 44

    var body: some View {
        HStack(spacing: 4) {
            ForEach(0..<figure.repeatCount, id: \.self) { _ in
                glyph
            }
        }
    }

    @ViewBuilder
    private var glyph: some View {
        let s = figure.repeatCount > 1 ? size * 0.62 : size
        switch figure.shape {
        case "square":
            shapeFill(RoundedRectangle(cornerRadius: 6, style: .continuous), s: s)
        case "circle", "dot":
            shapeFill(Circle(), s: s)
        case "arrow":
            ArrowShape()
                .rotation(.degrees(figure.rotationDegrees))
                .fill(Theme.accent)
                .frame(width: s, height: s)
        default: // triangle
            shapeFill(TriangleShape().rotation(.degrees(figure.rotationDegrees)), s: s)
        }
    }

    @ViewBuilder
    private func shapeFill<S: Shape>(_ shape: S, s: CGFloat) -> some View {
        if figure.isFilled {
            shape.fill(Theme.accent).frame(width: s, height: s)
        } else {
            shape.stroke(Theme.accent, lineWidth: 2.5).frame(width: s, height: s)
        }
    }
}

// MARK: - Shapes

/// 위를 향하는 정삼각형.
struct TriangleShape: Shape {
    func path(in rect: CGRect) -> Path {
        var p = Path()
        p.move(to: CGPoint(x: rect.midX, y: rect.minY))
        p.addLine(to: CGPoint(x: rect.maxX, y: rect.maxY))
        p.addLine(to: CGPoint(x: rect.minX, y: rect.maxY))
        p.closeSubpath()
        return p
    }
}

/// 위를 향하는 화살표(머리+기둥).
struct ArrowShape: Shape {
    func path(in rect: CGRect) -> Path {
        let w = rect.width, h = rect.height
        var p = Path()
        // 화살촉
        p.move(to: CGPoint(x: rect.midX, y: rect.minY))
        p.addLine(to: CGPoint(x: rect.minX, y: rect.minY + h * 0.5))
        p.addLine(to: CGPoint(x: rect.minX + w * 0.28, y: rect.minY + h * 0.5))
        // 기둥
        p.addLine(to: CGPoint(x: rect.minX + w * 0.28, y: rect.maxY))
        p.addLine(to: CGPoint(x: rect.minX + w * 0.72, y: rect.maxY))
        p.addLine(to: CGPoint(x: rect.minX + w * 0.72, y: rect.minY + h * 0.5))
        p.addLine(to: CGPoint(x: rect.maxX, y: rect.minY + h * 0.5))
        p.closeSubpath()
        return p
    }
}

// MARK: - Util

private extension Array {
    subscript(safe index: Int) -> Element? {
        indices.contains(index) ? self[index] : nil
    }
}

File: Android/app/src/main/java/com/jiny/catchtherule/ui/play/FigureView.kt
package com.jiny.catchtherule.ui.play

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.RowScope
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.graphics.drawscope.DrawScope
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.graphics.drawscope.rotate
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.jiny.catchtherule.core.model.Figure
import com.jiny.catchtherule.core.model.Puzzle
import com.jiny.catchtherule.ui.theme.AppColors
import com.jiny.catchtherule.ui.theme.card
import kotlin.math.cos
import kotlin.math.sin

// 도형 규칙 렌더러 — iOS FigureView.swift 와 동등.

// MARK: 숫자형 — 도형 + 숫자 슬롯

private const val FIG_SIDE = 150f   // dp

/** 숫자형 퍼즐: 완성된 예시 도형 + 빈칸 도형을 나란히 보여줘 규칙을 추론하게 한다. */
@Composable
fun FigureNumberRow(figures: List<Figure>, blankText: String = "?", feedback: AnswerFeedback? = null) {
    val n = figures.size.coerceAtLeast(1)
    val side = (if (n >= 3) 100f else if (n == 2) 138f else 150f).dp
    val spacing = if (n >= 3) 6.dp else 12.dp
    Row(
        Modifier.fillMaxWidth(),
        horizontalArrangement = Arrangement.spacedBy(spacing, Alignment.CenterHorizontally),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        figures.forEach { fig -> FigureNumberView(fig, blankText, feedback, side) }
    }
}

/** 숫자형 도형 1개(삼각형/사각형/원). 빈칸 슬롯은 blankText 로 표시. */
@Composable
fun FigureNumberView(figure: Figure, blankText: String = "?", feedback: AnswerFeedback? = null, side: Dp = FIG_SIDE.dp) {
    val slots = figure.slots ?: emptyList()
    val labelSide = maxOf(28.dp, side * 0.27f)
    val labelFont = maxOf(14f, side.value * 0.135f).sp
    Box(Modifier.size(side)) {
        // 윤곽선 + 원 세그먼트 구분선
        Canvas(Modifier.fillMaxSize()) {
            val inset = size.minDimension * 0.16f
            val strokePx = 2.dp.toPx()
            when (figure.shape) {
                "square" -> drawRoundRect(
                    color = AppColors.Stroke,
                    topLeft = Offset(inset, inset),
                    size = androidx.compose.ui.geometry.Size(size.width - inset * 2, size.height - inset * 2),
                    cornerRadius = androidx.compose.ui.geometry.CornerRadius(10.dp.toPx()),
                    style = Stroke(width = strokePx),
                )
                "circle" -> {
                    val r = size.minDimension / 2 - inset
                    val c = Offset(size.width / 2, size.height / 2)
                    drawCircle(AppColors.Stroke, radius = r, center = c, style = Stroke(width = strokePx))
                    // 세그먼트 경계 구분선
                    val n = slots.size.coerceAtLeast(1)
                    for (i in 0 until n) {
                        val deg = -90.0 + (i + 0.5) * 360.0 / n
                        val a = Math.toRadians(deg)
                        drawLine(
                            AppColors.Stroke.copy(alpha = 0.6f),
                            start = c,
                            end = Offset(c.x + (cos(a) * r).toFloat(), c.y + (sin(a) * r).toFloat()),
                            strokeWidth = 1.dp.toPx(),
                        )
                    }
                }
                else -> { // triangle
                    val p = Path().apply {
                        moveTo(size.width / 2, inset)
                        lineTo(size.width - inset, size.height - inset)
                        lineTo(inset, size.height - inset)
                        close()
                    }
                    drawPath(p, AppColors.Stroke, style = Stroke(width = strokePx))
                }
            }
        }
        // 슬롯 숫자 라벨
        val anchors = anchorFractions(figure.shape, slots.size)
        slots.forEachIndexed { i, value ->
            val (fx, fy) = anchors[i]
            SlotLabel(
                value = value,
                blankText = blankText,
                feedback = feedback,
                labelSide = labelSide,
                fontSize = labelFont,
                modifier = Modifier.offset(x = side * fx - labelSide / 2, y = side * fy - labelSide / 2),
            )
        }
    }
}

@Composable
private fun SlotLabel(
    value: String?,
    blankText: String,
    feedback: AnswerFeedback?,
    labelSide: Dp,
    fontSize: androidx.compose.ui.unit.TextUnit,
    modifier: Modifier,
) {
    val isBlank = value == null
    val strokeBrush: Brush = when {
        !isBlank -> SolidColor(AppColors.Stroke)
        feedback == AnswerFeedback.Correct -> SolidColor(AppColors.Success)
        feedback == AnswerFeedback.Wrong -> SolidColor(AppColors.Danger)
        else -> AppColors.AccentGradient
    }
    Box(
        modifier
            .size(labelSide)
            .clip(CircleShape)
            .background(if (isBlank) Color.White.copy(alpha = 0.05f) else AppColors.Card)
            .border(if (isBlank) 2.dp else 1.dp, strokeBrush, CircleShape),
        contentAlignment = Alignment.Center,
    ) {
        Text(
            text = if (isBlank) blankText else value.orEmpty(),
            color = if (isBlank) AppColors.TextPrimary else AppColors.TextSecondary,
            fontSize = fontSize, fontWeight = FontWeight.Bold, maxLines = 1,
        )
    }
}

/** 슬롯 앵커(0..1 비율, 시계방향, 위에서 시작). 마지막 1칸이 남으면 중앙(triangle/square). */
private fun anchorFractions(shape: String, n: Int): List<Pair<Float, Float>> {
    val center = 0.5f to 0.5f
    return when (shape) {
        "square" -> {
            val corners = listOf(0.16f to 0.16f, 0.84f to 0.16f, 0.84f to 0.84f, 0.16f to 0.84f)
            if (n > 4) corners + center else corners.take(n)
        }
        "circle" -> (0 until n).map { i ->
            val a = Math.toRadians(-90.0 + i * 360.0 / n.coerceAtLeast(1))
            (0.5f + (cos(a) * 0.2).toFloat()) to (0.5f + (sin(a) * 0.2).toFloat())
        }
        else -> { // triangle
            val verts = listOf(0.5f to 0.18f, 0.80f to 0.80f, 0.20f to 0.80f)
            // 삼각형 무게중심은 박스 정중앙(0.5)보다 아래 → 중앙 슬롯을 꼭짓점 평균으로.
            val triCenter = 0.5f to (verts.sumOf { it.second.toDouble() } / verts.size).toFloat()
            if (n > 3) verts + triCenter else verts.take(n)
        }
    }
}

// MARK: 시각형 — 순수 모양

/** 도형이 칸별로 변하는 시퀀스. null 셀이 빈칸(보기에서 고를 도형 자리). */
@Composable
fun FigureRow(figs: List<Figure?>, puzzle: Puzzle, reveal: Boolean, feedback: AnswerFeedback?, popScale: Float) {
    val count = figs.size.coerceAtLeast(1)
    val spacing = if (count >= 6) 6.dp else 10.dp
    val glyph = if (count >= 6) 34.dp else 42.dp
    Row(
        Modifier.fillMaxWidth(),
        horizontalArrangement = Arrangement.spacedBy(spacing),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        figs.forEach { fig ->
            if (fig != null) {
                Box(Modifier.weight(1f).height(76.dp).card(16.dp), contentAlignment = Alignment.Center) {
                    FigureGlyph(figure = fig, size = glyph)
                }
            } else {
                FigureBlankCell(puzzle, reveal, feedback, glyph)
            }
        }
    }
}

@Composable
private fun RowScope.FigureBlankCell(puzzle: Puzzle, reveal: Boolean, feedback: AnswerFeedback?, glyph: Dp) {
    val answerFig = if (reveal) puzzle.figureChoices?.firstOrNull { it.code == puzzle.answer } else null
    val strokeBrush: Brush = when (feedback) {
        AnswerFeedback.Correct -> SolidColor(AppColors.Success)
        AnswerFeedback.Wrong -> SolidColor(AppColors.Danger)
        else -> AppColors.AccentGradient
    }
    Box(
        Modifier
            .weight(1f)
            .height(76.dp)
            .clip(RoundedCornerShape(16.dp))
            .background(Color.White.copy(alpha = 0.03f))
            .border(2.dp, strokeBrush, RoundedCornerShape(16.dp)),
        contentAlignment = Alignment.Center,
    ) {
        if (answerFig != null) {
            FigureGlyph(figure = answerFig, size = glyph)
        } else {
            Text("?", color = AppColors.TextPrimary, fontSize = 30.sp, fontWeight = FontWeight.Bold)
        }
    }
}

/** 시각 규칙용 도형(회전/채움/개수). 시퀀스 셀·보기 공용. */
@Composable
fun FigureGlyph(figure: Figure, size: Dp) {
    val s = if (figure.repeatCount > 1) size * 0.62f else size
    Row(horizontalArrangement = Arrangement.spacedBy(4.dp), verticalAlignment = Alignment.CenterVertically) {
        repeat(figure.repeatCount) {
            Canvas(Modifier.size(s)) { drawGlyph(figure) }
        }
    }
}

private fun DrawScope.drawGlyph(figure: Figure) {
    val color = AppColors.Accent
    val filled = figure.isFilled
    val stroke = Stroke(width = 2.5.dp.toPx())
    val w = size.width
    val h = size.height
    rotate(figure.rotationDegrees) {
        when (figure.shape) {
            "square" -> if (filled)
                drawRoundRect(color, cornerRadius = androidx.compose.ui.geometry.CornerRadius(6.dp.toPx()))
            else drawRoundRect(color, cornerRadius = androidx.compose.ui.geometry.CornerRadius(6.dp.toPx()), style = stroke)
            "circle", "dot" -> if (filled)
                drawCircle(color, radius = w / 2)
            else drawCircle(color, radius = w / 2 - stroke.width / 2, style = stroke)
            "arrow" -> drawPath(arrowPath(w, h), color)
            else -> { // triangle
                val p = Path().apply {
                    moveTo(w / 2, 0f); lineTo(w, h); lineTo(0f, h); close()
                }
                if (filled) drawPath(p, color) else drawPath(p, color, style = stroke)
            }
        }
    }
}

private fun arrowPath(w: Float, h: Float): Path = Path().apply {
    moveTo(w / 2, 0f)
    lineTo(0f, h * 0.5f)
    lineTo(w * 0.28f, h * 0.5f)
    lineTo(w * 0.28f, h)
    lineTo(w * 0.72f, h)
    lineTo(w * 0.72f, h * 0.5f)
    lineTo(w, h * 0.5f)
    close()
}
